File: src/Models/Invoice.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace DashAccountingSystemV2.Models
{
    public class Invoice
    {
        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Computed)]
        public Guid Id { get; set; }

        [Required]
        public Guid TenantId { get; set; }
        public Tenant Tenant { get; private set; }

        [Required]
        public uint InvoiceNumber { get; set; }

        [Required]
        public InvoiceStatus Status { get; set; }

        [Required]
        public Guid CustomerId { get; set; }
        public Customer Customer { get; private set; }

        [EmailAddress]
        [MaxLength(256)]
        public string CustomerEmail { get; set; }

        [MaxLength(2048)]
        public string CustomerAddress { get; set; }

        [Required]
        public Guid InvoiceTermsId { get; set; }
        public InvoiceTerms InvoiceTerms { get; private set; }

        [Required]
        public DateTime IssueDate { get; set; }

        [Required]
        public DateTime DueDate { get; set; }

        [MaxLength(2048)]
        public string Message { get; set; }

        public ICollection<InvoiceLineItem> LineItems { get; set; } = new List<InvoiceLineItem>();

        public decimal Total
        {
            get
            {
                return LineItems?.Sum(li => li.Total) ?? 0.0m;
            }
        }

        public Guid? PaymentId { get; set; }
        public Payment Payment { get; private set; }

        [Required]
        [DatabaseGenerated(DatabaseGeneratedOption.Computed)]
        public DateTime Created { get; private set; }

        [Required]
        public Guid CreatedById { get; set; }
        public ApplicationUser CreatedBy { get; private set; }

        public DateTime? Updated { get; set; }

        public Guid? UpdatedById { get; set; }
        public ApplicationUser UpdatedBy { get; private set; }
    }
}

File: src/ClientApp/src/models/Invoice.ts
import Amount from './Amount';
import CustomerLite from './CustomerLite';
import InvoiceLineItem from './InvoiceLineItem';
import InvoiceStatus from './InvoiceStatus';
import InvoiceTerms from './InvoiceTerms';

export default interface Invoice {
    id?: string; // GUID (not required to create; required to update)
    tenantId: string; // GUID (required to create)
    invoiceNumber?: number; // uint (not required to create; required to update)
    status: InvoiceStatus; // required to create
    customerId: string | null; // GUID (required to create)
    customer?: CustomerLite; // not required to create; part of response
    customerAddress: string | null;
    customerEmail: string | null;
    invoiceTermsId: string | null; // GUID (required to create)
    invoiceTerms?: InvoiceTerms; // not required to create; part of response
    issueDate: string | null; // Date in YYYY-MM-DD format; required to create
    dueDate: string | null; // Date in YYYY-MM-DD format; required to create
    amount?: Amount; // not required to create; part of response
    message: string | null;
    lineItems: InvoiceLineItem[];
}
